slip-0019.md

```diff
@@ -5,24 +5,182 @@ Number:  SLIP-0019
+
+1. Ensure that bits 1 through 7 of *flags* are clear.
+2. Ensure that the wallet controls the private key to the provided *scriptPubKey*. This is typically done by using the provided BIP32 derivation path.
+3. If bit 0 (user confirmation) of *flags* is set, then prompt the user to confirm generation of the ownership proof with the given *commitmentData*. If the user does not confirm, then abort.
+4. Compute the ownership identifier for the scriptPubKey.
+5. Compile the *proofBody* and *proofFooter*, and generate the *proofSignature*.
+6. Return the *proofBody* and *proofSignature*.
+
+### Multi-signature scriptPubKeys
+
+The construction of a proof of ownership for a *m*-of-*n* multi-signature scriptPubKey requires a signing coordinator, i.e. a watch-only software wallet. The signing coordinator is assumed to have obtained the ownership identifiers of all *n* co-owners in advance. These ownership identifiers should generally be produced at the time of the creation of the multi-signature address.
+
+When constructing a proof of ownership, the signing coordinator prepares the *proofBody* and *proofFooter* and sends these to each signer together with any other required metadata, such as the BIP32 derivation path for the input. Each of the *m* signers then takes the following steps:
+
+1. Parse the *proofBody* and *proofFooter*. If the *version* is not recognized or if any of the bits 1 through 7 of *flags* is set, then abort.
+2. Derive the ownership identifier using the *scriptPubKey* provided in the *proofFooter*.
+3. If the derived ownership identifier is not listed in the *proofBody*, then abort.
+4. If bit 0 (user confirmation) of *flags* is set, then prompt the user to confirm generation of the ownership proof with the given *commitmentData*. If the user does not confirm, then abort.
+5. Return the signature for the provided *proofBody* and *proofFooter*.
+
+The signing coordinator collects all the signatures and combines them into a `SignatureProof` container to finalize the proof.
+
+## Proof usage
+
+### Verifying non-ownership of transaction inputs
+
+When a wallet is requested to sign a transaction, each external input SHOULD be accompanied with a proof of ownership so that the wallet may ascertain non-ownership of such an input in order to correctly inform the user about the amount they are spending in the transaction. For each external input the wallet takes the following steps:
+
+1. By reliable means obtain the scriptPubKey of the UTXO being spent by that input. Prior to SegWit version 1 witness programs this step involves acquiring the full transaction being spent and verifying its hash against that which is given in the outpoint.
+2. Parse the *proofBody*. If the *version* is not recognized or if any of the bits 1 through 7 of *flags* is set, then abort.
+3. Verify that the *proofSignature* is valid using the obtained scriptPubKey.
```

Comment:
probably should re-mention that validation is done via bip322

Reply:
Done in 2f84a7a.